File: packages/engine/src/core/digest-cache.ts
import { mkdirSync, readFileSync } from 'node:fs'
import { join } from 'node:path'
import { digestState } from '../projections/templates/digest-state'
import { writeFileAtomic } from './atomic'
import type { InitiativeState } from './fold'
import { ensureIndexDir, indexDir, logStat } from './index-store'
import { currentVersion, sortKeysDeep } from './snapshot'

/**
 * The session-start digest's state, cached per record (rust-core 4.4): the
 * digestState cut of the fold, which renders the same digest as the full state
 * (test/digest-state.test.ts). A hit renders without folding. At team scale
 * the fold was ~60% of session-start.
 *
 * KEY: the log's size and mtimeMs plus the engine version and schema hash,
 * exactly like the statusline facts. Every other session-start input (the
 * session id, git, repo memory, neighbours, rules, notices, env switches) is
 * still read live and handed to renderStatus unchanged.
 *
 * Derived and disposable in .sofar/.index/digest/<slug>.json. A miss, a
 * corrupt file or a mis-shaped one folds and rewrites. After the fold the log
 * is re-stat'd, so bytes that landed mid-fold are never cached under the old
 * key. The state is written as compact, key-sorted JSON, the form both
 * implementations produce byte for byte.
 */

const DIGEST_DIR = 'digest'
export const DIGEST_CACHE_VERSION = 3

interface DigestFile {
  v: number
  engine: string
  schema: string
  size: number
  mtimeMs: number
  state: InitiativeState
}

const isRecord = (v: unknown): v is Record<string, unknown> => typeof v === 'object' && v !== null && !Array.isArray(v)

/**
 * The cached value is trusted only in the shape renderStatus walks. The file
 * is ours and versioned, so this guards against truncation and foreign edits,
 * not against a different engine, which the key already refuses.
 */
function isDigestState(v: unknown): v is InitiativeState {
  if (!isRecord(v)) return false
  if (typeof v.slug !== 'string' || typeof v.goal !== 'string' || typeof v.status !== 'string') return false
  for (const key of ['phases', 'decisions', 'sessions', 'memories', 'runs', 'status_overrides', 'files_touched'] as const) {
    if (!Array.isArray(v[key])) return false
  }
  return isRecord(v.current) && isRecord(v.freshness) && isRecord(v.task_files) && (v.task_tests === undefined || isRecord(v.task_tests))
}

function digestFile(sofarDir: string, slug: string): string {
  return join(indexDir(sofarDir), DIGEST_DIR, `${slug}.json`)
}

/**
 * The digest state for `slug`: from the cache when its key still matches the
 * log, else digestState(fold()), written back. A missing log is never cached.
 */
export function cachedDigestState(
  sofarDir: string,
  slug: string,
  logPath: string,
  fold: () => InitiativeState,
): InitiativeState {
  const stat = logStat(logPath)
  if (stat === null) return digestState(fold())
  const { engine, schema } = currentVersion()
  const path = digestFile(sofarDir, slug)
  try {
    const raw = JSON.parse(readFileSync(path, 'utf8')) as Partial<DigestFile>
    if (
      raw.v === DIGEST_CACHE_VERSION &&
      raw.engine === engine &&
      raw.schema === schema &&
      raw.size === stat.size &&
      raw.mtimeMs === stat.mtimeMs &&
      isDigestState(raw.state)
    ) {
      return raw.state
    }
  } catch {
    // no file, or an unreadable one: a miss
  }
  const state = digestState(fold())
  const after = logStat(logPath)
  if (after !== null && after.size === stat.size && after.mtimeMs === stat.mtimeMs) {
    try {
      ensureIndexDir(sofarDir)
      mkdirSync(join(indexDir(sofarDir), DIGEST_DIR), { recursive: true })
      const file: DigestFile = { v: DIGEST_CACHE_VERSION, engine, schema, size: stat.size, mtimeMs: stat.mtimeMs, state }
      writeFileAtomic(path, `${JSON.stringify(sortKeysDeep(file))}\n`)
    } catch {
      // a cache that cannot be written is a miss next time, never an error
    }
  }
  return state
}

File: crates/sofar-core/src/digest_cache.rs
//! The session-start digest's state per record (`core/digest-cache.ts`,
//! rust-core 4.4): the [`digest_state`] cut, cached in the derived index under
//! the log's size and mtimeMs plus the engine version and schema hash, as
//! compact key-sorted JSON. A hit renders the digest without folding; a miss,
//! a corrupt file or a mis-shaped one folds and rewrites. The file is shared
//! with the TypeScript engine byte for byte.

use std::path::{Path, PathBuf};

use crate::append::fold_state;
use crate::atomic::write_file_atomic;
use crate::digest_state::digest_state;
use crate::fold::InitiativeState;
use crate::index_store::{LogStat, log_stat};
use crate::json::{self, Json, Object};
use crate::layout::Layout;
use crate::snapshot::current_version;

const DIGEST_DIR: &str = "digest";
const DIGEST_CACHE_VERSION: f64 = 2.0;

fn digest_file(layout: &Layout, slug: &str) -> PathBuf {
    layout
        .index_dir()
        .join(DIGEST_DIR)
        .join(format!("{slug}.json"))
}

#[allow(
    clippy::float_cmp,
    reason = "exact equality of a stored stat IS the contract"
)]
fn key_matches(o: &Object, engine: &str, schema: &str, stat: LogStat) -> bool {
    #[allow(
        clippy::cast_precision_loss,
        reason = "log sizes fit f64 exactly below 2^53"
    )]
    let size = stat.size as f64;
    o.get("v").and_then(Json::as_f64) == Some(DIGEST_CACHE_VERSION)
        && o.get("engine").and_then(Json::as_str) == Some(engine)
        && o.get("schema").and_then(Json::as_str) == Some(schema)
        && o.get("size").and_then(Json::as_f64) == Some(size)
        && o.get("mtimeMs").and_then(Json::as_f64) == Some(stat.mtime_ms)
}

/// `cachedDigestState`: from the cache when its key still matches the log,
/// else `digest_state(fold)`, written back. A missing log is never cached.
#[must_use]
pub fn cached_digest_state(layout: &Layout, slug: &str) -> InitiativeState {
    let log = layout.events_path(slug);
    let Some(stat) = log_stat(&log) else {
        return digest_state(&fold_state(layout, slug));
    };
    let version = current_version();
    let path = digest_file(layout, slug);
    if let Ok(bytes) = std::fs::read(&path)
        && let Ok(Json::Obj(raw)) = json::parse_bytes_fast(&bytes)
        && key_matches(&raw, &version.engine, &version.schema, stat)
        && let Some(state) = raw
            .get("state")
            .and_then(Json::as_obj)
            .and_then(InitiativeState::from_json)
    {
        return state;
    }
    let state = digest_state(&fold_state(layout, slug));
    // Re-stat after the fold: a write that landed while it ran must not be
    // cached under the key of bytes the fold never saw.
    if log_stat(&log) == Some(stat) {
        let _ = write_digest(
            layout,
            &path,
            &version.engine,
            &version.schema,
            stat,
            &state,
        );
    }
    state
}

fn write_digest(
    layout: &Layout,
    path: &Path,
    engine: &str,
    schema: &str,
    stat: LogStat,
    state: &InitiativeState,
) -> std::io::Result<()> {
    layout.ensure_index_dir()?;
    if let Some(dir) = path.parent() {
        std::fs::create_dir_all(dir)?;
    }
    #[allow(
        clippy::cast_precision_loss,
        reason = "log sizes fit f64 exactly below 2^53"
    )]
    let size = stat.size as f64;
    let mut o = Object::with_capacity(6);
    o.insert("v", Json::Num(DIGEST_CACHE_VERSION));
    o.insert("engine", Json::Str(engine.to_owned()));
    o.insert("schema", Json::Str(schema.to_owned()));
    o.insert("size", Json::Num(size));
    o.insert("mtimeMs", Json::Num(stat.mtime_ms));
    o.insert("state", state.to_json());
    let mut text = json::stringify_canonical(&Json::Obj(o));
    text.push('\n');
    write_file_atomic(path, text.as_bytes())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn repo() -> Layout {
        let dir = crate::testing::scratch_dir("digest-cache");
        let layout = Layout::new(&dir);
        let log = layout.events_path("x");
        std::fs::create_dir_all(log.parent().unwrap()).unwrap();
        let line = |id: &str, session: &str, kind: &str, payload: &str| {
            format!(
                "{{\"v\":1,\"id\":\"{id}\",\"ts\":\"2026-09-23T00:00:00.000Z\",\"initiative\":\"x\",\"session\":\"{session}\",\"source\":\"claude-code\",\"actor\":\"agent\",\"type\":\"{kind}\",\"payload\":{payload}}}\n"
            )
        };
        let text = [
            line(
                "01M00000000000000000000001",
                "a",
                "session_started",
                r#"{"tool":"t"}"#,
            ),
            line(
                "01M00000000000000000000002",
                "a",
                "session_ended",
                r#"{"summary":"s","next_action":"n"}"#,
            ),
        ]
        .concat();
        std::fs::write(&log, text).unwrap();
        layout
    }

    #[test]
    fn a_miss_writes_a_hit_reads_and_a_corrupt_file_is_a_miss() {
        let layout = repo();
        let want = digest_state(&fold_state(&layout, "x"));
        assert_eq!(cached_digest_state(&layout, "x"), want, "miss");
        let path = digest_file(&layout, "x");
        let good = std::fs::read_to_string(&path).expect("written on a miss");
        assert_eq!(cached_digest_state(&layout, "x"), want, "hit");
        for bad in [
            "nope".to_owned(),
            good.replace("\"v\":2", "\"v\":3"),
            good.replace("\"sessions\":[", "\"sessions\":7,\"x\":["),
        ] {
            assert_ne!(bad, good);
            std::fs::write(&path, &bad).unwrap();
            assert_eq!(cached_digest_state(&layout, "x"), want, "{bad}");
        }
    }
}
